core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -195,7 +195,18 @@ protected void perform(RelOptRuleCall call, Filter filter,
     // Try to push down filters in ON clause. A ON clause filter can only be
     // pushed down if it does not affect the non-matching set, i.e. it is
     // not on the side which is preserved.
-    if (RelOptUtil.classifyFilters(
+
+    // Anti-join on conditions can not be pushed into left or right, e.g. for plan:
+    //
+    //     Join(condition=[AND(cond1, $2)], joinType=[anti])
+    //     :  - prj(f0=[$0], f1=[$1], f2=[$2])
+    //     :  - prj(f0=[$0])
+    //
+    // The semantic would change if join condition $2 is pushed into left,
```

Comment:
I think it would be nice, if possible, to have a unit test showing this scenario where a wrong result is returned unless this patch is applied.

Reply:
Added, thanks